server/server.go

```diff
@@ -179,24 +179,29 @@ func (s *server) buildInvokeChain() {
 
 // RegisterService should not be called by users directly.
 func (s *server) RegisterService(svcInfo *serviceinfo.ServiceInfo, handler interface{}) error {
-	if s.svcInfo != nil {
-		panic(fmt.Sprintf("Service[%s] is already defined", s.svcInfo.ServiceName))
+	if svcInfo == nil {
+		return errors.New("svcInfo is nil. please specify non-nil svcInfo")
 	}
-	s.svcInfo = svcInfo
-	s.handler = handler
-	diagnosis.RegisterProbeFunc(s.opt.DebugService, diagnosis.ServiceInfoKey, diagnosis.WrapAsProbeFunc(s.svcInfo))
+	if handler == nil || reflect.ValueOf(handler).IsNil() {
+		return errors.New("handler is nil. please specify non-nil handler")
+	}
+	if s.svcMap[svcInfo.ServiceName] != nil {
+		panic(fmt.Sprintf("Service[%s] is already defined", svcInfo.ServiceName))
+	}
+	s.Lock()
```

Comment:
Here all svcMap operation is in single goroutine, is it necessary to add a locker for svcMap?

Reply:
@jayantxie Is it a single gorountine? I feel like users could call RegisterService in multiple goroutines.